### ISSUE TITLE
spring-boot with jsps in tomcat does not run on cloudfoundry

### ISSUE THREAD

**Issue by imod** [spring-projects/spring-boot#6657]
While upgrading from spring boot 1.3 to 1.4, we noticed that all our apps except one work. The one failing is the only one still using JSPs. 
Therefore I tried the `spring-boot-sample-tomcat-jsp` and was able to reproduce the issue: It fails  with the following stacktrace when deployed to cloudfoundry.

```
2016-08-15T21:22:12.90+0200 [APP/0]      OUT   .   ____          _            __ _ _
2016-08-15T21:22:12.90+0200 [APP/0]      OUT  /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
2016-08-15T21:22:12.90+0200 [APP/0]      OUT ( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
2016-08-15T21:22:12.90+0200 [APP/0]      OUT  \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
2016-08-15T21:22:12.90+0200 [APP/0]      OUT   '  |____| .__|_| |_|_| |_\__, | / / / /
2016-08-15T21:22:12.90+0200 [APP/0]      OUT  =========|_|==============|___/=/_/_/_/
2016-08-15T21:22:12.91+0200 [APP/0]      OUT  :: Spring Boot ::  (v1.4.1.BUILD-SNAPSHOT)
2016-08-15T21:22:13.15+0200 [APP/0]      OUT 2016-08-15 19:22:13.129  INFO 18 --- [           main] pertySourceApplicationContextInitializer : Adding 'cloud' PropertySource to ApplicationContext
2016-08-15T21:22:13.38+0200 [APP/0]      OUT 2016-08-15 19:22:13.387  INFO 18 --- [           main] nfigurationApplicationContextInitializer : Adding cloud service auto-reconfiguration to ApplicationContext
2016-08-15T21:22:13.43+0200 [APP/0]      OUT 2016-08-15 19:22:13.433  INFO 18 --- [           main] s.tomcat.jsp.SampleTomcatJspApplication  : Starting SampleTomcatJspApplication on ora15jmsr6k with PID 18 (/home/vcap/app/WEB-INF/classes started by vcap in /home/vcap/app)
2016-08-15T21:22:13.43+0200 [APP/0]      OUT 2016-08-15 19:22:13.437  INFO 18 --- [           main] s.tomcat.jsp.SampleTomcatJspApplication  : The following profiles are active: cloud
2016-08-15T21:22:13.65+0200 [APP/0]      OUT 2016-08-15 19:22:13.658  INFO 18 --- [           main] ationConfigEmbeddedWebApplicationContext : Refreshing org.springframework.boot.context.embedded.AnnotationConfigEmbeddedWebApplicationContext@5010be6: startup date [Mon Aug 15 19:22:13 UTC 2016]; root of context hierarchy
2016-08-15T21:22:15.47+0200 [APP/0]      OUT 2016-08-15 19:22:15.472  INFO 18 --- [           main] urceCloudServiceBeanFactoryPostProcessor : Auto-reconfiguring beans of type javax.sql.DataSource
2016-08-15T21:22:15.47+0200 [APP/0]      OUT 2016-08-15 19:22:15.475  INFO 18 --- [           main] urceCloudServiceBeanFactoryPostProcessor : No beans of type javax.sql.DataSource found. Skipping auto-reconfiguration.
2016-08-15T21:22:16.48+0200 [APP/0]      OUT 2016-08-15 19:22:16.480  INFO 18 --- [           main] s.b.c.e.t.TomcatEmbeddedServletContainer : Tomcat initialized with port(s): 8080 (http)
2016-08-15T21:22:16.49+0200 [APP/0]      OUT 2016-08-15 19:22:16.499  INFO 18 --- [           main] o.apache.catalina.core.StandardService   : Starting service Tomcat
2016-08-15T21:22:16.50+0200 [APP/0]      OUT 2016-08-15 19:22:16.500  INFO 18 --- [           main] org.apache.catalina.core.StandardEngine  : Starting Servlet Engine: Apache Tomcat/8.5.4
2016-08-15T21:22:16.87+0200 [APP/0]      OUT java.util.concurrent.ExecutionException: org.apache.catalina.LifecycleException: Failed to start component [StandardEngine[Tomcat].StandardHost[localhost].StandardContext[]]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	at java.util.concurrent.FutureTask.report(FutureTask.java:122) [na:1.8.0_91]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	at java.util.concurrent.FutureTask.get(FutureTask.java:192) [na:1.8.0_91]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:911) ~[tomcat-embed-core-8.5.4.jar!/:8.5.4]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	at org.apache.catalina.core.StandardHost.startInternal(StandardHost.java:890) [tomcat-embed-core-8.5.4.jar!/:8.5.4]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:152) [tomcat-embed-core-8.5.4.jar!/:8.5.4]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1403) [tomcat-embed-core-8.5.4.jar!/:8.5.4]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1393) [tomcat-embed-core-8.5.4.jar!/:8.5.4]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_91]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_91]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_91]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:158) [tomcat-embed-core-8.5.4.jar!/:8.5.4]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	... 6 common frames omitted
2016-08-15T21:22:16.87+0200 [APP/0]      OUT Caused by: java.lang.NullPointerException: null
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	at org.apache.tomcat.util.scan.StandardJarScanner.process(StandardJarScanner.java:325) ~[tomcat-embed-core-8.5.4.jar!/:8.5.4]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	at org.apache.tomcat.util.scan.StandardJarScanner.scan(StandardJarScanner.java:182) ~[tomcat-embed-core-8.5.4.jar!/:8.5.4]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	at org.apache.jasper.servlet.TldScanner.scanJars(TldScanner.java:262) ~[tomcat-embed-jasper-8.5.4.jar!/:8.5.4]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	at org.apache.jasper.servlet.TldScanner.scan(TldScanner.java:104) ~[tomcat-embed-jasper-8.5.4.jar!/:8.5.4]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	at org.apache.jasper.servlet.JasperInitializer.onStartup(JasperInitializer.java:103) ~[tomcat-embed-jasper-8.5.4.jar!/:8.5.4]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:152) [tomcat-embed-core-8.5.4.jar!/:8.5.4]
2016-08-15T21:22:16.87+0200 [APP/0]      OUT 	... 6 common frames omitted
```

a downgrade to `tomcat 8.0.36` does fix the issue

**Comment by wilkinsona** [comment]
Can you share the app or a sample that reproduces the problem? Failing that, the output from running with `logging.level.org.apache.tomcat.util.scan.StandardJarScanner` set to `trace` might let us figure out what's going on.

**Comment by wilkinsona** [comment]
Also posted on Stack Overflow: http://stackoverflow.com/questions/38961177/why-does-my-spring-boot-app-first-fail-and-then-suddenly-start-in-cloudfoundry

@imod Please don't cross-post without making it clear that you've done so

**Comment by wilkinsona** [comment]
Here's the log output from running `spring-boot-sample-web-jsp`:

```
2016-08-15T20:50:23.12+0100 [APP/0]      OUT 2016-08-15 19:50:23.125 DEBUG 19 --- [ost-startStop-1] o.a.tomcat.util.scan.StandardJarScanner  : Scanning /WEB-INF/lib for JARs
2016-08-15T20:50:23.13+0100 [APP/0]      OUT 2016-08-15 19:50:23.130 DEBUG 19 --- [ost-startStop-1] o.a.tomcat.util.scan.StandardJarScanner  : Not scanning JAR [/WEB-INF/lib/spring-context-4.3.2.RELEASE.jar] from /WEB-INF/lib
2016-08-15T20:50:23.13+0100 [APP/0]      OUT 2016-08-15 19:50:23.130 DEBUG 19 --- [ost-startStop-1] o.a.tomcat.util.scan.StandardJarScanner  : Not scanning JAR [/WEB-INF/lib/logback-classic-1.1.7.jar] from /WEB-INF/lib
2016-08-15T20:50:23.13+0100 [APP/0]      OUT 2016-08-15 19:50:23.130 DEBUG 19 --- [ost-startStop-1] o.a.tomcat.util.scan.StandardJarScanner  : Not scanning JAR [/WEB-INF/lib/jackson-databind-2.7.6.jar] from /WEB-INF/lib
2016-08-15T20:50:23.13+0100 [APP/0]      OUT 2016-08-15 19:50:23.130 DEBUG 19 --- [ost-startStop-1] o.a.tomcat.util.scan.StandardJarScanner  : Not scanning JAR [/WEB-INF/lib/spring-boot-starter-web-1.4.1.BUILD-SNAPSHOT.jar] from /WEB-INF/lib
2016-08-15T20:50:23.13+0100 [APP/0]      OUT 2016-08-15 19:50:23.131 DEBUG 19 --- [ost-startStop-1] o.a.tomcat.util.scan.StandardJarScanner  : Not scanning JAR [/WEB-INF/lib/jcl-over-slf4j-1.7.21.jar] from /WEB-INF/lib
2016-08-15T20:50:23.13+0100 [APP/0]      OUT 2016-08-15 19:50:23.131 DEBUG 19 --- [ost-startStop-1] o.a.tomcat.util.scan.StandardJarScanner  : Not scanning JAR [/WEB-INF/lib/spring-expression-4.3.2.RELEASE.jar] from /WEB-INF/lib
2016-08-15T20:50:23.13+0100 [APP/0]      OUT 2016-08-15 19:50:23.131 DEBUG 19 --- [ost-startStop-1] o.a.tomcat.util.scan.StandardJarScanner  : Scanning JAR [/WEB-INF/lib/spring-webmvc-4.3.2.RELEASE.jar] from /WEB-INF/lib
2016-08-15T20:50:23.13+0100 [APP/0]      OUT 2016-08-15 19:50:23.132 DEBUG 19 --- [ost-startStop-1] o.a.tomcat.util.scan.StandardJarScanner  : Scanning JAR at URL [file:/home/vcap/app/WEB-INF/lib/spring-webmvc-4.3.2.RELEASE.jar]
2016-08-15T20:50:23.19+0100 [APP/0]      OUT 2016-08-15 19:50:23.191 DEBUG 19 --- [ost-startStop-1] o.a.tomcat.util.scan.StandardJarScanner  : Not scanning JAR [/WEB-INF/lib/spring-aop-4.3.2.RELEASE.jar] from /WEB-INF/lib
2016-08-15T20:50:23.19+0100 [APP/0]      OUT 2016-08-15 19:50:23.191 DEBUG 19 --- [ost-startStop-1] o.a.tomcat.util.scan.StandardJarScanner  : Scanning JAR [/WEB-INF/lib/container_customizer-1.0.0_RELEASE.jar] from /WEB-INF/lib
2016-08-15T20:50:23.19+0100 [APP/0]      OUT 2016-08-15 19:50:23.191 DEBUG 19 --- [ost-startStop-1] o.a.tomcat.util.scan.StandardJarScanner  : Scanning JAR at URL [null]
```

**Comment by wilkinsona** [comment]
`Scanning JAR [/WEB-INF/lib/container_customizer-1.0.0_RELEASE.jar] from /WEB-INF/lib` logs a path returned from a call to `ServletContext.getResourcePaths(/WEB-INF/libs)`. `Scanning JAR at URL [null]` logs the URL returned from a call to `ServletContext.getResource(/WEB-INF/lib/container_customizer-1.0.0_RELEASE.jar)`.

I don't yet understand why a path returned by `getResourcePaths` cannot then be turned into a URL. It's also interesting that that jar that appears to be causing the problem is one that's added to the app by CloudFoundry /cc @nebhale

**Comment by wilkinsona** [comment]
With Tomcat 8.0.36 the path for the container customiser jar is successfully turned into a URL:

```
2016-08-15T21:05:45.38+0100 [APP/0]      OUT 2016-08-15 20:05:45.383 DEBUG 18 --- [ost-startStop-1] o.a.tomcat.util.scan.StandardJarScanner  : Scanning JAR [/WEB-INF/lib/container_customizer-1.0.0_RELEASE.jar] from /WEB-INF/lib
2016-08-15T21:05:45.38+0100 [APP/0]      OUT 2016-08-15 20:05:45.384 DEBUG 18 --- [ost-startStop-1] o.a.tomcat.util.scan.StandardJarScanner  : Scanning JAR at URL [file:/home/vcap/app/WEB-INF/lib/container_customizer-1.0.0_RELEASE.jar
```

**Comment by nebhale** [comment]
That JAR is going to be symlinked into the `WEB-INF/lib` directory.  It's raison d'etre is to [configure the Tomcat container to read symlinked JARs](https://github.com/cloudfoundry/java-buildpack-container-customizer/blob/master/src/main/java/org/cloudfoundry/container/customizer/AllowLinkingEmbeddedServletContainerCustomizer.java#L47).

**Comment by wilkinsona** [comment]
Thanks, @nebhale. Armed with that info and the customiser downloaded from https://java-buildpack.cloudfoundry.org/container-customizer/container-customizer-1.0.0_RELEASE.jar I can reproduce the problem locally

**Comment by nebhale** [comment]
@wilkinsona Sounds good. Let me know if we need to make any changes to our stuff.

**Comment by wilkinsona** [comment]
The problem's related to the caching of resources that's performed by `StandardRoot`.

The `/WEB-INF/lib/container-customizer-1.0.0_RELEASE.jar` resource is requested both before and after the container customizer has run. When it's requested before the customiser runs it cannot be found as linking is not enabled. When it's requested after the customiser runs, whether or not it's found depends on how quickly it's requested. The TTL of the cache introduces some latency to allowing linking having any effect. Stepping through in the debugger slows things down enough for the TTL to expire so that allow linking becomes effective and the resource can be found. Typically, during normal running things happen too quickly and the resource is not found.

**Comment by wilkinsona** [comment]
It works with 8.0.36 because it doesn't cache anything in `/WEB-INF/lib/":

```java
private boolean noCache(String path) {
    // Don't cache resources used by the class loader (it has its own cache)
    if (path.startsWith("/WEB-INF/classes/") ||
            path.startsWith("/WEB-INF/lib/")) {
        return true;
    }
    return false;
}
```

It fails in 8.5.4 because it doesn't cache anything in `/WEB-INF/lib/` where the path ends in `.class`:

```java
private boolean noCache(String path) {
    // Don't cache classes. The class loader handles this.
    if (path.endsWith(".class") &&
            (path.startsWith("/WEB-INF/classes/") ||
                    path.startsWith("/WEB-INF/lib/"))) {
        return true;
    }
    return false;
}
```

**Comment by wilkinsona** [comment]
@markt-asf Can you please take a look at the above? Is my analysis reasonable?

**Comment by wilkinsona** [comment]
The change in behaviour was introduced in [this commit](https://svn.apache.org/viewvc?view=revision&revision=1731275).

**Comment by markt-asf** [comment]
I don't see a Tomcat issue here although I'm prepared to be convinced otherwise.

My expectation is that no attempt is made to access the JAR before it and the symlink is in place. It looks as is the web app's resources are being changed while it is running. In that use case you are going to have to take account of any caching, including not found results.

**Comment by wilkinsona** [comment]
The jar and symlink making it appear to be in `WEB-INF/lib` are in place before the application is started. The only change that is made at runtime is a call to `StandardRoot.setAllowLinking(true)`. By the time this call has been made, previous requests for `/WEB-INF/lib/container_customizer-1.0.0_RELEASE.jar` have cached that it's not there (because allow linking was false). This caching only happens in 8.5. Once allow linking has been set to true (and no other change has been made) it takes up to the cache's TTL for it to notice that the resource is actually there.

In short, given the change to the caching behaviour in 8.5 that I linked to above, I think a call to `setAllowLinking(true)` needs to invalidate the cache. Without this cache invalidation, Tomcat 8.5.4 doesn't behave the same as 8.0.36.

**Comment by wilkinsona** [comment]
@markt-asf Any thoughts on my comment above? Have I managed to convince you otherwise?

**Comment by markt-asf** [comment]
You've convinced me to take a look at what can be done. One of the things Tomcat is not as good at as it could be is making clear what can safely changed dynamically and what can only be changed when the relevant component is stopped.

**Comment by markt-asf** [comment]
The fix was easy. Fixed for 9.0.0.M10, 8.5.5 and 8.0.37.

**Comment by wilkinsona** [comment]
Thanks, @markt-asf. Is there an ETA for 8.5.5?

**Comment by markt-asf** [comment]
I was aiming for early August. I'm still dealing with the backlog of stuff that built up while I was on leave in July and then at SpringOne Platform. Best guess right now is early Sept.

**Comment by wilkinsona** [comment]
That should work nicely. 1.4.1 is planned for 15 September.

**Comment by wilkinsona** [comment]
8.5.5 has fixed the problem. Its output is now the same as 8.0.36:

```
2016-09-14 14:57:32.299 DEBUG 52437 --- [ost-startStop-1] o.a.tomcat.util.scan.StandardJarScanner  : Scanning JAR [/WEB-INF/lib/container-customizer-1.0.0_RELEASE.jar] from /WEB-INF/lib
2016-09-14 14:57:32.300 DEBUG 52437 --- [ost-startStop-1] o.a.tomcat.util.scan.StandardJarScanner  : Scanning JAR at URL [file:/Users/awilkinson/dev/spring/spring-boot/master/spring-boot-samples/spring-boot-sample-tomcat-jsp/target/unpacked/WEB-INF/lib/container-customizer-1.0.0_RELEASE.jar]
```
